pitivi.1
.\" Copyright (C) 2010 The GNOME Foundation
.\"
.\" This is free software; you may redistribute it and/or modify
.\" it under the terms of the GNU Lesser General Public License as
.\" published by the Free Software Foundation; either version 2.1 of the License,
.\" or (at your option) any later version.
.\"
.\" This is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU Lesser General Public License for more details.
.\"
.\"You should have received a copy of the GNU Lesser General Public License along
.\"with this program; if not, write to the Free Software Foundation, Inc.,
.\"59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
.TH pitivi 1 "2010\-12\-08" "GNOME"
.SH NAME
pitivi \- Non-linear video editor
.SH SYNOPSIS
.B pitivi
.RI [PROJECT_FILE]
.P
.B pitivi
.RI -i
.RI [-a]
.RI [MEDIA_FILE]...
.SH DESCRIPTION
.B pitivi
is a free, intuitive and featureful movie editor for the Linux desktop.
.P
.B pitivi
starts the video editor, optionally loading PROJECT_FILE. If no project is
given, pitivi creates a new project. Alternatively, when -i is specified,
arguments are treated as clips to be imported into the project. If -a is
specified, these clips will also be added to the end of the project timeline.
.SH OPTIONS
.TP
.B \-h, \-\-help
Show help message and exit
.TP
.B \-i, \-\-import
Import each MEDIA_FILE into the project
.TP
.B \-a, \-\-add-to-timeline
Add each MEDIA_FILE to timeline after importing
.TP
.B \-d, \-\-debug
Run pitivi in the Python Debugger
.TP
.B \-n, \-\-no-ui
Run pitivi with no gui
.SH AUTHOR
This manual page was written by Hicham HAOUARI <hicham@fedoraproject.org>,
for the Fedora project (but may be used by others).
.SH BUGS
Please report bugs to http://bugzilla.gnome.org/enter_bug.cgi?product=pitivi


NEWS
2010 September 16 - PiTiVi 0.13.5

 * periodic backup of the current project file
 * easy crossfading transitions of overlapping clips
 * better icons for link and group operations
 * new add keyframe button
 * fixed support for the missing plugins installer
 * improved support for pictures
 * various performance improvements (less conversions, faster linking)

2010 March 10 - PiTiVi 0.13.4

 * video mixing/transparency support
 * icon view in source list
 * smoother scrolling
 * modeless splitting
 * seek on click
 * faster waveforms
 * zoom slider
 * UI beautifications
 * Speed optimisations
 * dbus/hal dependency now optional
 * translated in 30 languages

2009 September 12 - PiTiVi 0.13.3

 * Fix rendering failures
 * UI beautifications
 * Switch to themeable ruler
 * Speed optimisations
 * Show the project name in the window title

2009 August 13 - PiTiVi 0.13.2

 * Undo/Redo support
 * Audio-mixing
 * Ripple/Roll edit
 * misc fixes

2009 May 27 - PiTiVi 0.13.1

 * core rewrite
 * multi-layered timeline
 * trimming features
 * audio waveforms and video thumbnails in timeline
 * picture support
 * New project file format support

2008 Dec 11 - PiTiVi 0.11.3

 * Remove usage of gobject as much as possible from non-ui components
 * Make smarter choices about audio/video sinks 
 * Fix issues with seeking in ruler/viewer
 * general pylint cleanup
 * Picture support in the timeline
 * Improve viewer for proper Display Aspect Ratio
 * Timeline : Unlink-ing sources is now possible
 * Fix some issues when using very long sources
 * SourceList now detachable from main window
 * Now requires GNonLin 0.10.10 and python >= 2.5

2008 Oct 15 - PiTiVi 0.11.2

 * New advanced timeline interface by Brandon Lewis (SoC)
 * Capture interface for webcams and network sources by Sarath Lakshman (SoC).
 * Simple Timeline is gone.
 * Project save/load now activated by default
 * Cutting/Trimming/Removing features added to advanced timeline.
 * Misc fixes and improvements

2007 Nov 18 - PiTiVi 0.11.1

 * More work on project save/load, use PITIVI_FILE_SUPPORT=1 to try it.
 * Be more flexible with encoding caps, should fix more issues when encoding.
 * Fixed issues for non icon-theme-spec compliant systems
 * i18n updates
 * Misc fixes

2007 Oct 14 - PiTiVi 0.11.0

 * Merging of all the 2007 Summer-Of-Code branch
 * Simple timeline improvements : trimming and moving of sources
 * Simple timeline improvement : volume change
 * Advanced timeline re-activated
 * Advanced timeline features : cutting and moving of sources
 * Plugin framework added
 * File save/load framework added, not activated yet
 * i18n support, translated in 16 languages

2007 May 30 - PiTiVi 0.10.3

 * Improvement of first time user experience.
 * Frame-by-frame seeking
 * UI improvements for usability
 * New graphics
 * i18n support, translated in 14 languages
 * Critical bugfixes
 * Unit tests
 * Advanced view disabled by default

2007 Jan 31 - PiTiVi 0.10.2

 * UI improvements for usability
 * decodebin2 support
 * Rendering improvements
 * i18n support, translated in 12 languages
 * Critical bugfixes

2006 May 23 - PiTiVi 0.10.1

 * More UI improvements for usability
 * Works with X11/Imagesink
 * Handles more formats
 * More startup checks

2006 Apr 26 - PiTiVi 0.10.0

 * Lots of UI improvements for usability
 * Better error handling
 * Many playback issues fixed
 * Works in non-gnome environments


pitivi/pitivi.py
# PiTiVi , Non-linear video editor
#
#       pitivi/pitivi.py
#
# Copyright (c) 2005, Edward Hervey <bilboed@bilboed.com>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this program; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place - Suite 330,
# Boston, MA 02111-1307, USA.

"""
Main application
"""
import os
import gobject
import gst
import check
import instance
import device
from pitivigstutils import patch_gst_python
from playground import PlayGround
from project import Project, file_is_project
from effects import Magician
from configure import APPNAME
from settings import GlobalSettings
from threads import ThreadMaster
from pluginmanager import PluginManager
from signalinterface import Signallable
import instance

from gettext import gettext as _

# FIXME : Speedup loading time
# Currently we load everything in one go
# It would be better if a minimalistic UI could start up ASAP, without loading
# anything gst-related or that could slow down startup.
# AND THEN load up the required parts.
# This will result in a much better end-user experience

# FIXME : maybe we should have subclasses for UI and CLI

class Pitivi(object, Signallable):
    """
    Pitivi's main class

    Signals
        void new-project-loading()
            Pitivi is attempting to load a new project
        void new-project-loaded (project)
            a new project has been loaded, and the UI should refresh it's views
            * project - the project which has been loaded
        void new-project-failed(reason, uri)
            a new project could not be created
            * reason - the reason for failure
            * uri - the uri which failed to load (or None)
        boolean closing-project(project)
            pitivi would like to close a project. handlers should return false
            if they do not want this project to close. by default, assumes
            true.
            This signal should only be used by classes that might want to abort
            the closing of a project.
            * project - the project Pitivi would like to close
        void project-closed(project)
            The project is closed, it will be freed when the callback returns.
            Classes should connect to this instance when they want to know that
            data related to that project is no longer going to be used.
            * project - the project closed
        shutdown
            used internally, do not catch this signals"""

    __signals__ = {
        "new-project-loading" : ["project"],
        "new-project-loaded" : ["project"],
        "closing-project" : ["project"],
        "project-closed" : ["project"],
        "new-project-failed" : ["reason", "uri"],
        "shutdown" : None
        }

    def __init__(self, filepath=None):
        """
        initialize pitivi with the command line arguments
        """
        gst.log("starting up pitivi...")

        # patch gst-python for new behaviours
        # FIXME : this shouldn't be in this class
        patch_gst_python()

        # store ourself in the instance global
        if instance.PiTiVi:
            raise RuntimeWarning(
                _("There is already a %s instance, inform developers")
                % APPNAME)
        instance.PiTiVi = self

        # get settings
        self._settings = GlobalSettings()
        self.threads = ThreadMaster()
        #self.screencast = False

        self.plugin_manager = PluginManager(
            self.settings.get_local_plugin_path(),
            self.settings.get_plugin_settings_path())

        self.playground = PlayGround()
        self._current = Project(_("New Project"))
        self.effects = Magician()

        self.deviceprobe = device.get_probe()

    ## properties

    def _get_settings(self):
        return self._settings

    def _set_settings(self, settings):
        self._settings = settings
        # FIXME : we could notify this
    settings = property(_get_settings, _set_settings,
                        doc="The project-wide output settings")

    def _get_current(self):
        return self._current

    def _set_current(self, project):
        self._current = project
        # FIXME : we could notify this
    current = property(_get_current, _set_current,
                       doc="The currently used Project")

    ## public methods

    def loadProject(self, uri=None, filepath=None):
        """ Load the given file through it's uri or filepath """
        gst.info("uri:%s, filepath:%s" % (uri, filepath))
        if not uri and not filepath:
            self.emit("new-project-failed", _("No location given."),
                uri)
            return
        if filepath:
            if not os.path.exists(filepath):
                self.emit("new-project-failed",
                          _("File does not exist"), filepath)
                return
            uri = "file://" + filepath
        # is the given filepath a valid pitivi project
        if not file_is_project(uri):
            self.emit("new-project-failed", _("Not a valid project file."),
                uri)
            return
        # if current project, try to close it
        if self._closeRunningProject():
            project = Project(uri=uri)
            self.emit("new-project-loading", project)
            try:
                project.load()
                self.current = project
                self.emit("new-project-loaded", self.current)
            except:
                self.current = None
                self.emit("new-project-failed",
                    _("There was an error loading the file."), uri)

    def _closeRunningProject(self):
        """ close the current project """
        gst.info("closing running project")
        if self.current:
            if self.current.hasUnsavedModifications():
                if not self.current.save():
                    return False
            if self.emit("closing-project", self.current) == False:
                return False
            self.playground.pause()
            self.emit("project-closed", self.current)
            self.current = None
        return True

    def newBlankProject(self):
        """ start up a new blank project """
        # if there's a running project we must close it
        if self._closeRunningProject():
            project = Project(_("New Project"))
            self.playground.pause()
            self.emit("new-project-loading", project)
            self.current = project
            self.emit("new-project-loaded", self.current)

    def shutdown(self):
        """ Close PiTiVi
        Returns True if PiTiVi was successfully closed, else False
        """
        gst.debug("shutting down")
        # we refuse to close if we're running a user interface and the user
        # doesn't want us to close the current project.
        if not self._closeRunningProject():
            gst.warning("Not closing since running project doesn't want to close")
            return False
        self.threads.stopAllThreads()
        self.playground.shutdown()
        instance.PiTiVi = None
        self.emit("shutdown")
        return True



class InteractivePitivi(Pitivi):
    """ Class for PiTiVi instances that provide user interaction """

    def __init__(self, filepath=None, mainloop=None, *args, **kwargs):
        Pitivi.__init__(self, filepath=None,
                        *args, **kwargs)
        self.mainloop = mainloop

        from ui.mainwindow import PitiviMainWindow
        self._gui = PitiviMainWindow(self)
        self._gui.load()
        self._gui.show()

        if filepath:
            self.loadProject(filepath=filepath)

    # properties

    def _get_mainloop(self):
        return self._mainloop

    def _set_mainloop(self, mainloop):
        if hasattr(self, "_mainloop"):
            if self._mainloop != None:
                raise Exception("Mainloop already set !")
        if mainloop == None:
            mainloop = gobject.MainLoop()
        self._mainloop = mainloop
    mainloop = property(_get_mainloop, _set_mainloop,
                        doc="The MainLoop running the program")

    @property
    def gui(self):
        """The user interface"""
        return self._gui


    # PiTiVi method overrides
    def shutdown(self):
        if Pitivi.shutdown(self):
            if self.mainloop:
                self.mainloop.quit()
            return True
        return False

    def run(self):
        if self.mainloop:
            self.mainloop.run()

def main(argv):
    """ Start PiTiVi ! """
    from optparse import OptionParser
    check.initial_checks()
    parser = OptionParser()
    (unused_options, args) = parser.parse_args(argv[1:])
    if len(args) > 0:
        ptv = InteractivePitivi(filepath=args[0])
    else:
        ptv = InteractivePitivi()
    ptv.run()
